### ISSUE TITLE
Mono native crash during startup does not exit with a nonzero code

### ISSUE THREAD

**Issue by Xyene** [mono/mono#14416]
## Steps to Reproduce

1. Create `Test.cs` with the following content:
```c#
using System;

namespace App
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello, World!");
        }
    }
}
```

2. Compile with `mcs Test.cs` to produce `Test.exe`.
3. `ulimit -Sv 28000`. Might need to vary the exact number to produce the erroneous behaviour.
4. Run `mono Test.exe`. This should cause the runtime to fail during initialization.

### Current Behavior
```
xyene@iorwerth:/code$ mono Test.exe
Could not allocate 8192 bytes
=================================================================
        Native Crash Reporting
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================
/proc/self/maps:
4030f000-4031f000 rwxp 00000000 00:00 0
40e91000-40ea1000 rwxp 00000000 00:00 0
55a981869000-55a981cc1000 r-xp 00000000 08:00 16535                      /usr/bin/mono-sgen
55a981ec1000-55a981ec7000 r--p 00458000 08:00 16535                      /usr/bin/mono-sgen
55a981ec7000-55a981ecd000 rw-p 0045e000 08:00 16535                      /usr/bin/mono-sgen
55a981ecd000-55a981ee4000 rw-p 00000000 00:00 0
55a983013000-55a9830d3000 rw-p 00000000 00:00 0                          [heap]
7fd29a355000-7fd29a5c9000 r--p 00000000 08:00 619998                     /usr/lib/mono/gac/System/4.0.0.0__b77a5c561934e089/System.dll
7fd29a5c9000-7fd29a5ca000 ---p 00000000 00:00 0
7fd29a5ca000-7fd29a5cb000 rw-p 00000000 00:00 0
7fd29a5cb000-7fd29a5d3000 ---p 00000000 00:00 0
7fd29a5d3000-7fd29a7ca000 rw-p 00000000 00:00 0
7fd29a7ca000-7fd29ad7d000 r-xp 00000000 08:00 619838                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7fd29ad7d000-7fd29af7c000 ---p 005b3000 08:00 619838                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7fd29af7c000-7fd29af7d000 r--p 005b2000 08:00 619838                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7fd29af7d000-7fd29af7e000 rw-p 005b3000 08:00 619838                     /usr/lib/mono/aot-cache/amd64/mscorlib.dll.so
7fd29af7e000-7fd29afa8000 rw-p 00000000 00:00 0
7fd29afa8000-7fd29b3ff000 r--p 00000000 08:00 619832                     /usr/lib/mono/4.5/mscorlib.dll
7fd29b3ff000-7fd29c3ff000 rw-p 00000000 00:00 0
7fd29c3ff000-7fd29c400000 ---p 00000000 00:00 0
7fd29c400000-7fd29d000000 rw-p 00000000 00:00 0
7fd29d03f000-7fd29d0fc000 rw-p 00000000 00:00 0
7fd29d0fd000-7fd29d101000 rw-p 00000000 00:00 0
7fd29d101000-7fd29d160000 ---p 00000000 00:00 0
7fd29d160000-7fd29d2f5000 r-xp 00000000 08:00 7248                       /lib/x86_64-linux-gnu/libc-2.24.so
=================================================================
        Basic Fault Adddress Reporting
=================================================================
Memory around native instruction pointer (0x7fd29d192fff):0x7fd29d192fef  d2 4c 89 ce bf 02 00 00 00 b8 0e 00 00 00 0f 05  .L..............
0x7fd29d192fff  48 81 c4 88 00 00 00 44 89 c0 5b c3 0f 1f 44 00  H......D..[...D.
0x7fd29d19300f  00 48 8b 15 61 5e 36 00 f7 d8 64 89 02 eb ca 66  .H..a^6...d....f
0x7fd29d19301f  90 85 ff 78 0c f7 df e9 e5 02 00 00 0f 1f 44 00  ...x..........D.
=================================================================
        Native stacktrace:
=================================================================
        0x55a98199c448 - mono : (null)
        0x55a98192f8f1 - mono : (null)
        0x7fd29d7270e0 - /lib/x86_64-linux-gnu/libpthread.so.0 : (null)
        0x7fd29d192fff - /lib/x86_64-linux-gnu/libc.so.6 : gsignal
        0x7fd29d19442a - /lib/x86_64-linux-gnu/libc.so.6 : abort
        0x55a981b86ed7 - mono : (null)
        0x55a981b6d4b1 - mono : (null)
        0x55a981b86dbb - mono : (null)
        0x55a981b871e3 - mono : monoeg_g_logv
        0x55a981b8729f - mono : monoeg_g_log
        0x55a981b86c46 - mono : monoeg_malloc
        0x55a981a68e0f - mono : mono_mempool_alloc
        0x55a981a68ef2 - mono : mono_mempool_alloc0
        0x55a9818bfc50 - mono : (null)
        0x55a9818e6ad7 - mono : (null)
        0x55a9818ce4b3 - mono : (null)
        0x55a9818e6ad7 - mono : (null)
        0x55a9818ce4b3 - mono : (null)
        0x55a9819a1116 - mono : (null)
        0x55a9819a3504 - mono : (null)
        0x55a9818ac601 - mono : (null)
        0x55a9818ad985 - mono : (null)
        0x55a981a8ae80 - mono : (null)
        0x55a981a91551 - mono : (null)
        0x55a9818ab35b - mono : mono_resolve_patch_target
        0x55a9819220fe - mono : (null)
        0x55a981922744 - mono : (null)
        0x55a981923f45 - mono : (null)
        0x55a981933cc5 - mono : (null)
        0x4030fb13 - Unknown
=================================================================
        Telemetry Dumper:
=================================================================
* Assertion: should not be reached at mono-state.c:270
xyene@iorwerth:/code$ echo $?
0
```

### Expected Behavior

Mono should be exiting with a nonzero exit code to communicate the startup failure. Minor bug: there is also a typo in "Basic Fault Adddress Reporting".

## On which platforms did you notice this

[ ] macOS
[x] Linux
[ ] Windows

**Version Used**:

```
$ mono --version
Mono JIT compiler version 5.18.1.0 (tarball Fri Mar 15 20:41:21 UTC 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          yes(600)
        Suspend:       preemptive
        GC:            sgen (concurrent by default)

$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 9.8 (stretch)
Release:        9.8
Codename:       stretch

$ uname -a
Linux localhost 4.18.16-x86_64-linode118 #1 SMP PREEMPT Mon Oct 29 15:38:25 UTC 2018 x86_64 GNU/Linux
```

---

I also reproduced this on Mono 5.20, with the error occurring elsewhere:

```
Could not allocate 10720 bytes

=================================================================
        Native Crash Reporting
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================

=================================================================
        Basic Fault Adddress Reporting
=================================================================
Memory around native instruction pointer (0x155554063fff):0x155554063fef  d2 4c 89 ce bf 02 00 00 00 b8 0e 00 00 00 0f 05  .L..............
0x155554063fff  48 81 c4 88 00 00 00 44 89 c0 5b c3 0f 1f 44 00  H......D..[...D.
0x15555406400f  00 48 8b 15 61 5e 36 00 f7 d8 64 89 02 eb ca 66  .H..a^6...d....f
0x15555406401f  90 85 ff 78 0c f7 df e9 e5 02 00 00 0f 1f 44 00  ...x..........D.

=================================================================
        Native stacktrace:
=================================================================
        0x555555688268 - mono : (null)
        0x55555561b071 - mono : (null)
        0x1555545f80e0 - /lib/x86_64-linux-gnu/libpthread.so.0 : (null)
        0x155554063fff - /lib/x86_64-linux-gnu/libc.so.6 : gsignal
        0x15555406542a - /lib/x86_64-linux-gnu/libc.so.6 : abort
        0x555555876527 - mono : (null)
        0x55555585caf1 - mono : (null)
        0x55555587640b - mono : (null)
        0x555555876833 - mono : monoeg_g_logv
        0x5555558768ef - mono : monoeg_g_log
        0x555555876296 - mono : monoeg_malloc
        0x55555575758c - mono : mono_mempool_alloc
        0x5555555e051a - mono : (null)
        0x55555568e1ab - mono : (null)
        0x55555568f324 - mono : (null)
        0x555555597851 - mono : (null)
        0x55555561e9c5 - mono : (null)
        0x55555561f3b3 - mono : (null)
        0x40000393 - Unknown

=================================================================
        Telemetry Dumper:
=================================================================
* Assertion: should not be reached at mini-posix.c:249
```

Exact Mono version details:

```
$ mono --version
Mono JIT compiler version 5.20.1.19 (tarball Thu Apr 11 09:07:22 UTC 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          yes(600)
        Suspend:       hybrid
        GC:            sgen (concurrent by default)
```

**Comment by marek-safar** [comment]
@alexanderkyte could you look into this one?

**Comment by alexanderkyte** [comment]
I think that you're running into an issue that has already been fixed in master @Xyene.

The crash you're seeing occurs when we cannot allocate the memory because the working directory isn't writable. The fallback in that eventuality was added here: https://github.com/mono/mono/pull/13447

We *do* need to mmap code though, or we should fail with an assertion. Based on the error I see there: 

```
Could not allocate 10720 bytes
```

which comes from one of these lines:

```
mono/eglib/gmem.c:      g_error ("Could not allocate %i bytes", size);
mono/eglib/gmem.c:      g_error ("Could not allocate %i bytes", x);
mono/eglib/gmem.c:      g_error ("Could not allocate %i (%i * %i) bytes", x*n, n, x);
mono/metadata/appdomain.c:              mono_error_set_out_of_memory (error, "Could not allocate %ud bytes to copy raw assembly data", raw_assembly_len);
mono/metadata/marshal.c:                mono_error_set_out_of_memory (error, "Could not allocate %" G_GSIZE_FORMAT " bytes", size);
mono/metadata/object.c:         mono_error_set_out_of_memory (error, "Could not allocate %i bytes", m_class_get_instance_size (klass));
mono/metadata/object.c:         mono_error_set_out_of_memory (error, "Could not allocate %i bytes", m_class_get_instance_size (klass));
mono/metadata/object.c:         mono_error_set_out_of_memory (error, "Could not allocate %i bytes", m_class_get_instance_size (vtable->klass));
mono/metadata/object.c:                         mono_error_set_out_of_memory (error, "Could not allocate %i bytes", MONO_ARRAY_MAX_SIZE);
mono/metadata/object.c:         mono_error_set_out_of_memory (error, "Could not allocate %i bytes", MONO_ARRAY_MAX_SIZE);
mono/metadata/object.c:                 mono_error_set_out_of_memory (error, "Could not allocate %i bytes", MONO_ARRAY_MAX_SIZE);
mono/metadata/object.c:                 mono_error_set_out_of_memory (error, "Could not allocate %i bytes", MONO_ARRAY_MAX_SIZE);
mono/metadata/object.c:         mono_error_set_out_of_memory (error, "Could not allocate %zd bytes", (gsize) byte_len);
mono/metadata/object.c:         mono_error_set_out_of_memory (error, "Could not allocate %i bytes", MONO_ARRAY_MAX_SIZE);
mono/metadata/object.c:         mono_error_set_out_of_memory (error, "Could not allocate %zd bytes", (gsize) byte_len);
mono/metadata/object.c:         mono_error_set_out_of_memory (error, "Could not allocate %i bytes", -1);
mono/metadata/object.c:         mono_error_set_out_of_memory (error, "Could not allocate %" G_GSIZE_FORMAT " bytes", size);
mono/metadata/object.c:         mono_error_set_out_of_memory (error, "Could not allocate %" G_GSIZE_FORMAT " bytes", size);
mono/mini/interp/interp.c:                                      mono_error_set_out_of_memory (error, "Could not allocate %i bytes", m_class_get_instance_size (vtable->klass));
mono/sgen/sgen-memory-governor.c:       fprintf (stderr, "Error: Garbage collector could not allocate %zu bytes of memory for %s.\n", requested_size, assert_description);
```

it seems likely that you're running into a situation that makes the mmap fail as well. If that's the case, the failure you see is the current planned way we have to error out. 

I'm going to confirm that this is what's happening on my end though. First, I will make a fix for https://github.com/mono/mono/issues/14469

**Comment by alexanderkyte** [comment]
I have confirmed that limiting the virtual memory causes the crash reporter to not be able to mmap some private memory, or to create the mmap'ed file we would otherwise use to hold the state. 

Things from this to fix:
- Nonzero exit code should be set.
- Typo around `Basic Fault Adddress Reporting`

**Comment by alexanderkyte** [comment]
Both of those have been fixed by #14540
